src/PAO_grid_transform_module.f90

```diff
@@ -115,127 +115,147 @@ subroutine single_PAO_to_grid(pao_fns)
     use pao_format
 
-    implicit none 
+    implicit none
     integer,intent(in) :: pao_fns
 
     !local
-    real(double):: dcellx_block,dcelly_block,dcellz_block
-    integer :: ipart,jpart,ind_part,ia,ii,icover,ig_atom
-    real(double):: xatom,yatom,zatom,alpha,step
-    real(double):: xblock,yblock,zblock
-    integer :: the_species
-    integer :: j,iblock,the_l,ipoint, igrid
-    real(double) :: r_from_i
-    real(double) :: rr,a,b,c,d,x,y,z,nl_potential
-    integer :: no_of_ib_ia, offset_position
-    integer :: position,iatom
-    integer :: stat, nl, npoint, ip, this_nsf
-    integer :: i,m, m1min, m1max,acz,m1,l1,count1
-    integer     , allocatable :: ip_store(:)
-    real(double), allocatable :: x_store(:)
-    real(double), allocatable :: y_store(:)
-    real(double), allocatable :: z_store(:)
-    real(double), allocatable :: r_store(:)
-    real(double) :: coulomb_energy
-    real(double) :: rcut
-    real(double) :: r1, r2, r3, r4, core_charge, gauss_charge
-    real(double) :: val, theta, phi, r_tmp
-    integer :: max_num_blocks, current_num_blocks
-    real(double), dimension(:), allocatable :: temp_block_storage
+    integer :: iblock,ia,ipart,ip,l1,acz,m1 ! Loop index variables
+    integer :: nblock, npart, natom ! Array dimensions
+    integer :: jpart,ind_part,ii,icover,ig_atom,iatom ! Temporary indices for calculating species
+    integer :: position,next_offset_position ! Temporary indices to gridfunctions%griddata
+    integer :: my_species ! Temporary variables to reduce indirect accesses
+    integer :: npoint ! outputs of check_block
+    integer :: count1 ! incremented counter, maps from (l1, acz, m1) to linear index of gridfunctions%griddata
+    real(double):: dcellx_block,dcelly_block,dcellz_block ! grid dimensions, should be moved
+    real(double) :: x,y,z ! Temporary variables to reduce indirect accesses
+    real(double) :: rcut ! Input to check_block
+    real(double) :: val ! output, written into gridfunctions%griddata
+    integer, allocatable, dimension(:) :: ip_store ! outputs of check_block
+    integer, allocatable, dimension(:,:,:) :: offset_position, species_store ! precomputed variables
+    real(double), allocatable, dimension(:) :: x_store, y_store, z_store, r_store ! outputs of check_block
+    real(double), allocatable, dimension(:) :: xblock,yblock,zblock ! inputs to check_block, precomputed
+    real(double), allocatable, dimension(:,:,:) :: xatom,yatom,zatom ! inputs to check_block, precomputed
+
+    nblock = domain%groups_on_node
+    npart = maxval(naba_atoms_of_blocks(atomf)%no_of_part)
+    natom = maxval(naba_atoms_of_blocks(atomf)%no_atom_on_part)
 
     call start_timer(tmr_std_basis)
     call start_timer(tmr_std_allocation)
-    allocate(ip_store(n_pts_in_block),x_store(n_pts_in_block),y_store(n_pts_in_block),z_store(n_pts_in_block), &
-         r_store(n_pts_in_block))
+    allocate(xblock(nblock), yblock(nblock), zblock(nblock))
+    allocate(xatom(natom, npart, nblock))
```

Comment:
xblock etc, xatom etc, species_store should not be arrays I think

Reply:
You are definitely right about `xblock`, `yblock`, `zblock`. We only need the species to calculate the offsets. I have moved those to the openmp loop. With `xatom`, `yatom`, `zatom` and `species_store`, it depends on whether we want to redo the calculation of `icover` and `ig_atom` in the parallel loop. Maybe that would make sense, it won't be that expensive and will save us memory allocations.